Add new PPD module for plots without `y` variable

jgabry opened the pull request:
Closes #151 
Advances #97
Advances #196

This is a big PR (sorry!) that adds a lot of new functions and refactors a bunch of existing code along the way. The bad news is that there are a ton of changed files, the good news is that many of them are generated .Rd files and svgs for vdiffr, although still many files with changed R code. @tjmahr Given how large this PR is, what would be the way that I could best help you go through this? For example, do you want me to generate and post images of all the new plots? Or schedule a time to talk through some of it? Anyway, totally up to you, just let me know what I can do to help. 

The major new addition is a new "module" PPD -- see `help("PPD-overview")` -- which is for plotting posterior or prior predictive distributions without comparing to any observed data. These functions have prefix `ppd_` instead of `ppc_` and take an argument `ypred` instead of `yrep` and `y`. 

Some additional refactoring (discussed in https://github.com/stan-dev/bayesplot/issues/151#issuecomment-496323557 and subsequent comments) makes it so that `_grouped()` functions no longer require duplicating much of the same code from the non-grouped function. Instead, the non-grouped function is called and then faceting is added. 

#### New PPD functions

PPD-distributions
  - ppd_data()
  - ppd_dens_overlay()
  - ppd_ecdf_overlay()
  - ppd_dens()
  - ppd_hist()
  - ppd_freqpoly()
  - ppd_freqpoly_grouped()
  - ppd_boxplot()

PPD-intervals
  - ppd_intervals_data()
  - ppd_intervals()
  - ppd_intervals_grouped()
  - ppd_ribbon()
  - ppd_ribbon_grouped()

PPD-test-statistics
  - ppd_stat_data()
  - ppd_stat()
  - ppd_stat_grouped()
  - ppd_stat_freqpoly()
  - ppd_stat_freqpoly_grouped()
  - ppd_stat_2d()

#### More ppc_*_data() functions

Also, it made sense while refactoring to add the following `ppc_*_data()` functions that were missing: 

  - ppc_bars_data()
  - ppc_error_binnned_data()
  - ppc_error_data()
  - ppc_scatter_avg_data()
  - ppc_scatter_data()
  - ppc_stat_data()

#### More functions get facet_args arguments that were missing

  - ppc_scatter()
  - ppc_scatter_avg_grouped()
   - ppc_error_hist()
   - ppc_error_hist_grouped()
   - ppc_error_scatter()
   - ppc_error_binned()

jgabry commented:
All checks now passing via GitHub actions

bnicenboim commented:
Hi, I wanted to check what's the status of this PR ;)
No rush, but I would love to get rid of `y` from the prior predictive checks in our teaching material. 
Is there something I can do to help?

bnicenboim commented:
ehem, yet again I taught an intro to Bayes and I had to deal with the y in prior predictive checks.
Is there something that will break if I just use this branch instead of master?

jgabry commented:
@bnicenboim Thanks for bringing up this PR in your [forum post](https://discourse.mc-stan.org/t/bayesplot-updates/24662/4?u=jonah), and sorry for the delay! 

I think originally @tjmahr was going to review this but it's a big PR and I'm guessing he's just been too busy. @tjmahr Are you still interested in looking at this PR (I think you already looked at some of the branch based on our earlier conversations in https://github.com/stan-dev/bayesplot/issues/151)? 

It also looks like there are some merge conflicts created by having merged other PRs. I'll take a look at those.

jgabry commented:
> It also looks like there are some merge conflicts created by having merged other PRs. I'll take a look at those.

I think I've got the merge conflicts fixed except for the ones in ppc-distributions.R, which might take a bit longer to sort out

jgabry commented:
> I think I've got the merge conflicts fixed except for the ones in ppc-distributions.R, which might take a bit longer to sort out

Ok sorted that out too. About to push the fixes but I think the tests will fail until I fix a few other things. But have to take a break for now

jgabry commented:
Ok another update: 

I think I fixed all the merge conflicts and I updated to using the new testthat/vdiffr `_snaps` folder instead of `figs` for the visual tests. All the tests now pass locally for me!

@bnicenboim Thanks again for pushing me to revisit this. If you want to try out this branch and make sure it works as expected that would be great! (I've done that but it would be good to have some more testers)

@tjmahr If you're still up for reviewing this I think it's ready (it's big but see my very first comment in this PR for some comments about that). If you're too busy (no worries!) let me know and I can try to find someone else. 

At this point I think we should merge this one before merging any of the other open PRs. We can update the other PR branches after this is merged. I think fixing any conflicts that way will be easier than if we do the reverse order.

bnicenboim commented:
I'm checking the branch now, and something that I don't understand is the need for different function names `ppd_*`. Why not just allow for y=NULL, or y =NA

tjmahr commented:
> I'm checking the branch now, and something that I don't understand is the need for different function names `ppd_*`. Why not just allow for `y=NULL`, or `y =NA` in the regular `ppc_*` functions?

The upside would be that there are fewer functions in the packages and related plots would draw on some shared code. The downsides would be

* The ppc functions have to work down two separate branches.
* Looking at just a posterior predictive distribution requires you to set a mandatory value to NA or NULL (so it's a weird interface).
* The abbreviation "posterior predictive check" is less meaningful.

tjmahr commented:
> I think originally @tjmahr was going to review this but it's a big PR and I'm guessing he's just been too busy. @tjmahr Are you still interested in looking at this PR (I think you already looked at some of the branch based on our earlier conversations in #151)?

Because this is so big, can I review in chunks?

tjmahr commented on NAMESPACE:
```diff
@@ -167,6 +197,7 @@ importFrom(dplyr,pull)
 importFrom(dplyr,rename)
 importFrom(dplyr,select)
 importFrom(dplyr,summarise)
+importFrom(dplyr,summarise_at)
```
> Scoped verbs (_if, _at, _all) have been superseded by the use of across() in an existing verb. See vignette("colwise") for details.

Something to keep an eye on.

jgabry replied:
Thanks for the heads up. I can't keep up with all these changes ;) I'll look into using across() instead.

jgabry replied:
switched to using across and pushed so will mark this as resolved

tjmahr commented on NAMESPACE:
```diff
@@ -157,6 +185,8 @@ import(stats)
 importFrom(dplyr,"%>%")
 importFrom(dplyr,arrange)
 importFrom(dplyr,count)
+importFrom(dplyr,everything)
```
This is a tidyselect function which dplyr reexports: https://tidyselect.r-lib.org/reference/everything.html

jgabry replied:
switched to tidyselect::everything and pushed so will mark this as resolved

tjmahr commented on R/helpers-ppc.R:
```diff
@@ -135,136 +172,103 @@ validate_x <- function(x = NULL, y, unique_x = FALSE) {
+#'   `called_from_internal=TRUE` which can be detected by the function to be
+#'   called so it knows not to warn about the `group` and `facet_args` arguments.
+#' @noRd
+ungroup_call <- function(fn, call) {
```
Huh, this is quite the strategy. The sight of it makes me nervous but it does yield shorter code.

jgabry replied:
Yeah, it's basically what I came up with to avoid duplicating code in both the grouped and ungrouped functions. I first proposed it in [this comment](https://github.com/stan-dev/bayesplot/issues/151#issuecomment-496323557) to which you rightfully [responded](https://github.com/stan-dev/bayesplot/issues/151#issuecomment-496885009) that it's "pretty tricky stuff" (which it is), but then it sounded like you eventually [agreed](https://github.com/stan-dev/bayesplot/issues/151#issuecomment-496999158) with the approach. 

I'm not opposed to a different approach if you have a better idea (there are probably better ideas out there). That would probably mean starting from scratch on this but I guess that's ok if there's a much better path.

jgabry replied:
I guess we should finish having this conversation before you go down the rabbit hole of reviewing the whole thing, just in case we decide to change the approach.

tjmahr replied:
Ha, it's funny that past me had the same response. I don't want to rehash anything, so let's keep it. It's indirect enough that it can be refactored out later if needed.

jgabry replied:
Ok cool. Yeah we can always refactor in the future if we want.

bnicenboim commented:
On Tue, Oct 5, 2021, 1:37 PM TJ Mahr ***@***.***> wrote:

> I'm checking the branch now, and something that I don't understand is the
> need for different function names ppd_*. Why not just allow for y=NULL,
> or y =NA in the regular ppc_* functions?
>
> The upside would be that there are fewer functions in the packages and
> related plots would draw on some shared code. The downsides would be
>
>    - The ppc functions have to work down two separate branches.
>
> But now all the ppc functions are duplicated, that also doesn't look ideal.

>
>    - Looking at just a posterior predictive distribution requires you to
>    set a mandatory value to NA or NULL (so it's a weird interface).
>
> Well,the first thing I tried when I wanted prior predictive distributions
was to try to set y = NULL, since I wanted identical plots but just without
y.
I would at least allow for the option to set y= NULL as another way to call
the PPD functions.

(A more disruptive option that will break others people code is to change
the order of y and yep, but I guess it's going to be problematic).

>
>    - The abbreviation "posterior predictive check" is less meaningful.
>
> I would read it as prior predictive checks :)

But also not using y is not only for prior predictions, one can also plot
predictive distributions for conditions where there is still not data. I
had an experiment with set sizes of 2, 4, and 6 and i used ppc functions to
generate predictions for set sizes that weren't present in the experiment
(3 and 5).


>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/bayesplot/pull/222#issuecomment-934328156>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNUQ6VX7VHKI4IHWDARGCTUFLPQVANCNFSM4N77CWFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jgabry commented:
> > I think originally @tjmahr was going to review this but it's a big PR and I'm guessing he's just been too busy. @tjmahr Are you still interested in looking at this PR (I think you already looked at some of the branch based on our earlier conversations in #151)?
> 
> Because this is so big, can I review in chunks?

Yeah it's huge (sorry). Whatever approach you want to take is fine by me!

jgabry commented:
Thanks for the comments @bnicenboim, much appreciated! 

> But now all the ppc functions are duplicated, that also doesn't look ideal.

Yeah some of the ppc functions get corresponding ppd functions. Some of them don't because things like calculating errors (for the `ppc_error_*` plots) don't make sense when there is no `y`.   

> - Looking at just a posterior predictive distribution requires you to set a mandatory value to NA or NULL (so it's a weird interface).

I agree with @tjmahr that it would be a weird interface. We can't change the order of the y and yrep arguments at this point, and I didn't like the idea of an interface requiring setting the first mandatory argument to to NA or NULL. But it's totally reasonable to disagree on that point. 

> I would at least allow for the option to set y= NULL as another way to call the PPD functions.

In general I prefer to have one way to do something instead of supporting two different ways to call the same function. It would be one thing if this provided a much easier way for the user to to code it, but the user having to changing the prefix from `ppc` to `ppd` doesn't seem like a bigger burden than having to set `y=NULL`. Both are just a few keystrokes. If we're going to support `y=NULL` then I would rather that just be the only option, although I still slightly prefer having the separate `ppd` functions. 

> - The abbreviation "posterior predictive check" is less meaningful.
> I would read it as prior predictive checks :)

But it's not just for prior predictive checks, also for plots of posterior predictions that aren't compared to observed data. So `ppc` isn't quite right to describe that. `ppd` works for either prior predictive distribution or posterior predictive distribution. 

> But also not using y is not only for prior predictions, one can also plot predictive distributions for conditions where there is still not data. I had an experiment with set sizes of 2, 4, and 6 and i used ppc functions to generate predictions for set sizes that weren't present in the experiment (3 and 5).

That's true but I think the existing functions are already suitable for that, right?

jgabry commented:
@tjmahr Just a heads up. I've been pushing more to this branch but it's just merging in master and fixing conflicts after I merge other PRs that have been waiting to get into master. I won't make any substantive changes to this branch while you're reviewing it.

jgabry commented:
> I pushed four commits today.

Awesome, thanks TJ! Those all look great. 

 > I am able to pass the testthat tests and rmdcheck::rcmdcheck() on my machine. When I tried to run the R package check through RStudio, I got the same error as the CI build:

I think maybe you're not getting the error when running `rmdcheck::rcmdcheck()` because the code in the mcmc diagnostics vignette is set to not run unless the environment variable `NOT_CRAN` is TRUE (which it is in the github actions workflows). This is so that, when submitting to CRAN, the vignette is rendered locally and sent to CRAN instead of generated by CRAN (at one point they complained that the rstan compilation in the visual diagnostics vignette took too long so this way avoids that issue while not giving up using rstan in the vignette). 

What happens if you run this? 

```r
withr::with_envvar(c("NOT_CRAN" = "true"), rmdcheck::rcmdcheck())
```

Does that pass or result in the same error? Either way, I think the error is actually an rstan problem and not a bayesplot problem because this

> argument "RMU" is missing, with no default

is a known rstan issue on windows that I think is fixed on rstan's github but maybe not released yet (I'm a bit behind on my rstan knowledge these days as I've transitioned to maintaining cmdstanr and other people are working on rstan). 

> We should get the CI builds working though. 

Definitely agree. I think since the CI issues are unrelated to this branch we can proceed with this PR but I can then start a new branch to sort out the CI issues. 

> I had to manually install hexbin, and ggfortify in order to get tests to work on my machine.

I'm guessing we're missing a `skip_if_not_installed("hexbin")`. I'll look into that.

tjmahr commented:
Okay, update the NEWS and merge!

bnicenboim commented:
Hi, yet again it's me :)
I wanted to check with @tjmahr  @jgabry, what the status of this branch is. It looked fine to me in linux, not sure what happens in Windows and Mac.

jgabry commented:
Working on this and getting ready to finally do the next release of bayesplot including these changes. Sorry for the delay! Will be merging this PR today hopefully, as soon as the checks pass (except for the Windows issue, which is not related to bayesplot and needs to be fixed in rstan).

jgabry commented:
Finally passing all checks! Merging now and will start preparing the release. After the release can start working on merging the other open PRs.

jgabry commented:
On CRAN now finally!
